tests/modules/thunder_test_mock_request/src/ThunderTestMockRequestServiceProvider.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Drupal\thunder_test_mock_request;
+
+use Drupal\Core\DependencyInjection\ContainerBuilder;
+use Drupal\Core\DependencyInjection\ServiceProviderBase;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Add mock handler and HTTP middleware.
+ */
+class ThunderTestMockRequestServiceProvider extends ServiceProviderBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function alter(ContainerBuilder $container) {
+    parent::alter($container);
+
+    $container->getDefinition('http_handler_stack')
+      ->addArgument(new Reference('thunder_test_mock_request.mock_handler'));
+
+    $container
```

Comment:
Conceptually I'd have this code above the other one. But also why do you need the other one. I would have hoped the tagged service was enough.

Reply:
We have to pass the MockHandler into the factory method:
```
  http_handler_stack:
    class: GuzzleHttp\HandlerStack
    public: false
    factory: GuzzleHttp\HandlerStack::create
    configurator: ['@http_handler_stack_configurator', configure]
```
Otherwise, the HandlerStack would be initialized with whatever \GuzzleHttp\choose_handler() returns.